### PR TITLE
Modified annotate_sex()

### PR DESCRIPTION
**Pull request by wlu04**

### CODE REVIEW

File: gnomad/sample_qc/pipeline.py [hunk 1]
```diff
@@ -269,9 +284,22 @@ def annotate_sex(
         mt = mt.filter_rows(
             (hl.len(mt.alleles) == 2) & hl.is_snp(mt.alleles[0], mt.alleles[1])
         )
-    mt = hl.filter_intervals(
-        mt, [hl.parse_locus_interval(contig) for contig in x_contigs]
-    )
+
+    if is_vds:
```
Comment (jkgoodrich) [inline gnomad/sample_qc/pipeline.py:288]: I think I'm not fully following this. For a VDS at this point `mt` is just the split biallelic variant data MT so `filter_intervals` should work fine on it right? And this only keeps the `mtds.variant_data`

File: gnomad/sample_qc/pipeline.py [hunk 2]
```diff
@@ -282,6 +310,9 @@ def annotate_sex(
         logger.info("Filtering to provided sites")
         mt = mt.annotate_rows(**sites_ht[mt.row_key])
         mt = mt.filter_rows(hl.is_defined(mt[aaf_expr]))
+        if is_vds:
```
Comment (jkgoodrich) [inline gnomad/sample_qc/pipeline.py:313]: Is this needed? If only keeping `mtds.variant_data` and not using `mtds` after this, I don't think `hl.vds.filter_variants` is needed

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
Some questions, but it's very possible I'm not following because I haven't worked with the VDS format as long as you. Thanks for putting in this PR!

File: gnomad/sample_qc/pipeline.py [hunk 3]
```diff
@@ -252,14 +252,29 @@ def annotate_sex(
...
+            calling_intervals=included_intervals,
+            normalization_contig=normalization_contig,
         )
-    else:
-        raise NotImplementedError(
-            "Imputing sex ploidy does not exist yet for dense data."
+        ploidy_ht = ploidy_ht.transmute(
+            chrX_ploidy=ploidy_ht.x_ploidy, chrY_ploidy=ploidy_ht.y_ploidy
         )
+        mtds = hl.vds.split_multi(mtds, filter_changed_loci=True)
```
[the remark this answers is not in the document]
Comment (wlu04) [reply gnomad/sample_qc/pipeline.py]: This is right, they work equivalently on `mtds.variant_data`. Sorry about that. And in this case, do we want to add a warning or something about using `filter_chaged_loci`  before loading the data?

[the remark this answers is not in the document]
Comment (jkgoodrich) [reply gnomad/sample_qc/pipeline.py]: No, I think it's fine without it, it's outside the scope of the function and the MT doesn't even need to be split before using it

File: gnomad/sample_qc/pipeline.py [hunk 4]
```diff
@@ -238,11 +239,12 @@ def annotate_sex(
         - Y_karyotype (str): Sample's chromosome Y karyotype.
         - sex_karyotype (str): Sample's sex karyotype.
 
-    :param mt: Input MatrixTable
+    :param mtds: Input MatrixTable or VariantDataset
     :param bool is_sparse: Whether input MatrixTable is in sparse data format
     :param excluded_intervals: Optional table of intervals to exclude from the computation.
     :param included_intervals: Optional table of intervals to use in the computation. REQUIRED for exomes.
     :param normalization_contig: Which chromosome to use to normalize sex chromosome coverage. Used in determining sex chromosome ploidies.
+    :param reference_genome: Reference genome used for constructing interval list. Default: 'GRCh38'
```
Comment (jkgoodrich) [inline gnomad/sample_qc/pipeline.py:247]: I wonder if we should just infer this from the MT? example: https://github.com/broadinstitute/gnomad_methods/blob/978b57762108d6ccbe979ed7a867ffe361d146e6/gnomad/utils/filtering.py#L133

File: gnomad/sample_qc/pipeline.py [hunk 5]
```diff
@@ -252,14 +254,32 @@ def annotate_sex(
...
     logger.info("Imputing sex chromosome ploidies...")
-    if is_sparse:
-        ploidy_ht = impute_sex_ploidy(
-            mt, excluded_intervals, included_intervals, normalization_contig
+
+    is_vds = isinstance(mtds, hl.vds.VariantDataset)
+    if is_vds:
+        if excluded_intervals is not None:
+            raise NotImplementedError(
+                "excluded_intervals is not used when imputing sex chromosome ploidy for VDS"
```
Comment (jkgoodrich) [inline gnomad/sample_qc/pipeline.py:262]:
```suggestion
                "The use of the parameter excluded_intervals is currently not implemented for imputing sex chromosome ploidy on a VDS"
```

**Review by jkgoodrich — CHANGES_REQUESTED** [review]
This is close to ready, just a couple more comments (minus the addition of `hl.vds.impute_sex_chromosome_ploidy` to Hail for tests to pass)

File: gnomad/sample_qc/pipeline.py [hunk 6]
```diff
@@ -225,7 +225,7 @@ def annotate_sex(
 
     Return Table with the following fields:
         - s (str): Sample
-        - chr20_mean_dp (float32): Sample's mean coverage over chromosome 20.
+        - chr20_mean_dp/autosomal_mean_dp (float32): Sample's mean coverage over chromosome 20.
```
Comment (jkgoodrich) [inline gnomad/sample_qc/pipeline.py:228]:
```suggestion
        - {normalization_contig}_mean_dp (float32): Sample's mean coverage over the specified `normalization_contig`.
```

File: gnomad/sample_qc/pipeline.py [hunk 7]
```diff
@@ -252,14 +252,37 @@ def annotate_sex(
...
     logger.info("Imputing sex chromosome ploidies...")
-    if is_sparse:
-        ploidy_ht = impute_sex_ploidy(
-            mt, excluded_intervals, included_intervals, normalization_contig
+
+    is_vds = isinstance(mtds, hl.vds.VariantDataset)
+    if is_vds:
+        if excluded_intervals is not None:
+            raise NotImplementedError(
+                "The use of the parameter excluded_intervals is currently not implemented for imputing sex chromosome ploidy on a VDS"
```
Comment (jkgoodrich) [inline gnomad/sample_qc/pipeline.py:260]:
```suggestion
                "The use of the parameter 'excluded_intervals' is currently not implemented for imputing sex chromosome ploidy on a VDS!"
```

File: gnomad/sample_qc/pipeline.py [hunk 8]
```diff
@@ -252,14 +252,37 @@ def annotate_sex(
...
-    else:
-        raise NotImplementedError(
-            "Imputing sex ploidy does not exist yet for dense data."
+        ploidy_ht = ploidy_ht.rename(
+            {
+                "x_ploidy": "chrX_ploidy",
+                "y_ploidy": "chrY_ploidy",
+                "x_mean_dp": "chrX_mean_dp",
+                "y_mean_dp": "chrY_mean_dp",
+            }
```
Comment (jkgoodrich) [inline gnomad/sample_qc/pipeline.py:273]:
```suggestion
                "autosomal_mean_dp": f"{normalization_contig}_mean_dp",
            }
```

**Review by jkgoodrich — APPROVED** [review]
two more tiny suggestions, and it looks like `hl.vds.impute_sex_chromosome_ploidy` might now be in Hail, so a recommit should hopefully fix that error

**Review by jkgoodrich — APPROVED** [review]
LGTM

**Review by jkgoodrich — APPROVED** [review]

**Pull request merged**